zcash_primitives/src/transaction/builder.rs

```diff
@@ -393,22 +512,325 @@ impl Progress {
     }
 }
 
+pub struct SaplingBuilder<P: consensus::Parameters> {
+    anchor: Option<bls12_381::Scalar>,
+    value_balance: Amount,
+    spends: Vec<SpendDescriptionInfo>,
+    outputs: Vec<SaplingOutput<P>>,
+    change_address: Option<(OutgoingViewingKey, PaymentAddress)>,
+}
+
+impl<P: consensus::Parameters> SaplingBuilder<P> {
+    fn new() -> Self {
+        SaplingBuilder {
+            anchor: None,
+            value_balance: Amount::zero(),
+            spends: vec![],
+            outputs: vec![],
+            change_address: None,
+        }
+    }
+
+    pub fn value_balance(&self) -> Amount {
+        self.value_balance
+    }
+
+    /// Adds a Sapling note to be spent in this transaction.
+    ///
+    /// Returns an error if the given Merkle path does not have the same anchor as the
+    /// paths for previous Sapling notes.
+    fn add_spend<R: RngCore>(
+        &mut self,
+        mut rng: R,
+        extsk: ExtendedSpendingKey,
+        diversifier: Diversifier,
+        note: Note,
+        merkle_path: MerklePath<Node>,
+    ) -> Result<(), Error> {
+        // Consistency check: all anchors must equal the first one
+        let cmu = Node::new(note.cmu().into());
+        if let Some(anchor) = self.anchor {
+            let path_root: bls12_381::Scalar = merkle_path.root(cmu).into();
+            if path_root != anchor {
+                return Err(Error::AnchorMismatch);
+            }
+        } else {
+            self.anchor = Some(merkle_path.root(cmu).into())
+        }
+
+        let alpha = jubjub::Fr::random(&mut rng);
+
+        self.value_balance += Amount::from_u64(note.value).map_err(|_| Error::InvalidAmount)?;
+
+        self.spends.push(SpendDescriptionInfo {
+            extsk,
+            diversifier,
+            note,
+            alpha,
+            merkle_path,
+        });
+
+        Ok(())
+    }
+
+    /// Adds a Sapling address to send funds to.
+    #[allow(clippy::too_many_arguments)]
+    pub fn add_output<R: RngCore>(
+        &mut self,
+        mut rng: R,
+        params: &P,
+        target_height: BlockHeight,
```

Comment:
I think these should be stored in the builder, not passed on every invocation, since these _must_ be the same for every output.